### DISCUSSION TITLE
Enabling microdot to run other things while app.run()

### DISCUSSION THREAD

**Discussion by AppsByDavideV** [miguelgrinberg/microdot#122] [Q&A]
Hi, 
great app! 
I would like to run other tasks while the server is running and serving pages, but my knowledge about micropython is too low (just started, coming from c/c++ arduino). I tried to play with uasyncio  (and even I  had a look at the code into microdot_asyncio  app.start_server) without lucky. 

Is there a way to run other tasks while the server is running?
 Thank you so much for answer ;)

**Comment by miguelgrinberg** [comment]
The most common way is to use asyncio, so you were in the right place. The `app.start_server` coroutine handles the server, you can start this task and others and they'll all share time under asyncio.

**Reply by AppsByDavideV** [reply]
I think I'm missing something... my (test) code is as follow. The server works, but still blocked . The prints 'server started' and 'running other tasks' never takes place....
Can you kindly have a look at this? Thanks! (sorry it looks not formatted correctly in this reply)

```python
import uasyncio
from microdot_asyncio import Microdot, Response, send_file
from led_module import LEDModule
from micropython import const
import time

#Consts
CHECK_TIME=const(1000)

app = Microdot()
Response.default_content_type = 'text/html'

# Our LED Module
led_module = LEDModule(48)


#@app.route('/')
#async def index(request):
#    return render_template('index.html', led_value=led_module.get_value())

@app.route('/')
async def index(request):
    print(request.path)
    return send_file('/static/html/index.html')


@app.route('/toggle')
async def toggle_led(request):
    print("Receive Toggle Request!")
    led_module.toggle()
    return "OK"


@app.route('/shutdown')
async def shutdown(request):
    request.app.shutdown()
    return 'The server is shutting down...'


@app.route('/static/<path:path>')
def static(request, path):
    if '..' in path:
        # directory traversal is not allowed
        return 'Not found', 404
    return send_file('static/' + path)

async def main():
    await app.start_server(port=80,debug=False)
    print('server started...')
    

uasyncio.run(main())

#running other task from here
print('Running other tasks...')
while True:
    time.sleep_ms(1)
```

**Reply by miguelgrinberg** [reply]
The problem is that you are awaiting the `start_server()` call. Basically you are telling asyncio to block until the server returns.

Try using `uasyncio.create_task(app.start_server(...))` instead. This will launch the server but will not wait for the server to return.

**Reply by AppsByDavideV** [reply]
tried this, but in this case the server doesn't work:

`async def main():
    #await app.start_server(port=80,debug=False)
    uasyncio.create_task(app.start_server(port=80,debug=False))
    print('server started...')
    

uasyncio.run(main())

#running other task from here
print('Running other tasks...')
while True:
    time.sleep_ms(1)`

**Reply by miguelgrinberg** [reply]
I recommend that you work through an asyncio or uasyncio tutorial to learn how to use it better.

Try something like this:

```python
async def main():
    uasyncio.create_task(app.start_server(port=80,debug=False))
    print('server started...')

    #running other task from here
    print('Running other tasks...')
    while True:
        time.sleep_ms(1000)

uasyncio.run(main())
```

**Reply by AppsByDavideV** [reply]
Hi, thanks.
Just tried the code above.
But it doesn't work either. It is not blocking but the server doesn't work. It looks like if the server is not looping itself, it can't serve pages.
I forgot to say the board is ESP32S3 16M flash and 8M psram  Micropython V1.19.
Have a nice day!

**Reply by miguelgrinberg** [reply]
I wrote the code above from memory, copy/pasting from your code, which now I realize that it is wrong in several ways.

You cannot mix sync and async code, as that causes some blocking. The sleep that you are using is blocking and cannot be used with uasyncio. You should instead use uasyncio.sleep. Let me state again that asyncio is not a trivial library to master, there is a lot of things that you need to know to be able to use it without creating issues.

**Reply by AppsByDavideV** [reply]
Hi, thanks again.
I will look into it. Anyway, from my last test, the only way to get the server working is using app.run() or await app.server_start(). Other combinations may start the server in 'not' blocking mode but it doesn't serve pages.

Will try again when I'll have time.
Great job!
\Davide

**Reply by miguelgrinberg** [reply]
The other combinations that you tried did not work because the other activity you were adding was blocking the server from running.

**Reply by AppsByDavideV** [reply]
Hi,
finally I've got it ;)

The code is like follow , just in case someone else needs it. Tested on ESP32-S3
Thank you
\Davide

```
# https://github.com/miguelgrinberg/microdot  https://microdot.readthedocs.io/en/latest/

import uasyncio
from microdot_asyncio import Microdot, Response, send_file
from led_module import LEDModule
from micropython import const
import time

#Consts
CHECK_TIME=const(1000)

app = Microdot()
Response.default_content_type = 'text/html'

# RGB LED Module
led_module = LEDModule(48)

@app.route('/')
async def index(request):
    print(request.path)
    return send_file('/static/html/index.html')


@app.route('/toggle')
async def toggle_led(request):
    print("Receive Toggle Request!")
    led_module.toggle()
    return "OK"


@app.route('/shutdown')
async def shutdown(request):
    request.app.shutdown()
    return 'The server is shutting down...'


@app.route('/static/<path:path>')
def static(request, path):
    if '..' in path:
        # directory traversal is not allowed
        return 'Not found', 404
    return send_file('static/' + path)
    
async def server():
    # start the server at port 80 (web)
    await app.start_server(port=80)

async def runForever():
    print ('Running loop...')
    checkTime=time.ticks_ms()
    #start of main loop forever
    while True:
        # execute task checkTime
        if time.ticks_diff(time.ticks_ms(),checkTime) > CHECK_TIME:
            # time expired, do something
            print ('Time: ' + str(time.ticks_ms()))
            checkTime=time.ticks_ms()
        # other tasks
        #
        await uasyncio.sleep_ms(1)


async def main():
    uasyncio.create_task(server())
    uasyncio.create_task(runForever())
    while True:
        await uasyncio.sleep_ms(1)	# time after that the device stops running    
    
# start async server and other tasks
# https://docs.micropython.org/en/latest/library/uasyncio.html?highlight=uasyncio
uasyncio.run(main())	# this may end or never end, depending on started tasks
```

**Reply by OhanaCode** [reply]
Thank you all... this solution(s) is what I was in desperate need of!

**Reply by maxi07** [reply]
Thank you so much for the example code @AppsByDavideV!

**Reply by themeadery** [reply]
@AppsByDavideV Is there a reason this line of code:
`def static(request, path):`
 is not another "async def" like the rest? Ex:
 `async def static(request, path):`

**Reply by AppsByDavideV** [reply]
Hi, you are right, it is **_async_** like other defs. Probably a leftover during copy/paste of the code.
Bye

**Reply by jouellnyc** [reply]
You are both gentlemen and scholars!

**Reply by jouellnyc** [reply]
In case this is obvious, I am having trouble installing microdot_asyncio.

I can't find it - I've looked in the repo here.
I can't upip it as upip is not on my pico:
"MicroPython v1.22.1 on 2024-01-05; Raspberry Pi Pico W with RP2040 (installed via Thonny 4.1.2)"
I chased upip and upip_tar files manualy to get upip back,  but was not successful after failing to find zlib.
I understand there is discussion re: mip but @AppsByDavideV , how did you install it?

**Reply by miguelgrinberg** [reply]
@jouellnyc this discussion started before the release of version 2 of Microdot. Back in version 1 there was a thing called `microdot_asyncio`. In version 2, Microdot is always asynchronous, so you get asyncio support without having to install any add-ons. As to how to install it, do not use upip, see the documentation to learn which files from this repository you need to copy to your device.

**Reply by AppsByDavideV** [reply]
Hi @jouellnyc , for readability you can create a folder called 'lib' into your device and copy the files inside it.
@miguelgrinberg already answered to your question, anyway the related files you need are in the repo, into  'src/microdot' folder, but this is a new version of the library.
Bye

**Reply by jouellnyc** [reply]
Aha. Yes. Thanks @miguelgrinberg! I had used microdot a year+ back with excellent results but fumbled w/the asyncio as I am getting back up to speed. It is developer's like you creating these create products that make working with micro controllers so much fun.

**Comment by Jarshvor** [comment]
Hi, and sorry to hijack this discussion, but I feel it might be related:

I am wondering if it is possible to run Microdot (have the server running) while at the same time retain access to rshell or some form of repl prompt. I would like to be able to keep interacting with the rshell after the server has started, either via app.server_start() or app.run().

I have read a bit on asyncio, and tried poking about with it, (also tried AppsByDavideV's code above).
I am quite new to micropython, microdot, rshell etc so I am not entirely sure if this is a limitation on the rshell/repl side and how that's started.

Any sort of guidance or assistance would be greatly appreciated (even if to say "nah, what you want isnt really possible").
And thank you for creating Microdot. =)

**Reply by miguelgrinberg** [reply]
I have not tested this myself, but I think the aiorepl package will allow you to do what you want: https://github.com/micropython/micropython-lib/tree/master/micropython/aiorepl

**Reply by Jarshvor** [reply]
Thanks for the super fast reply Miguel!
I was indeed able to get it working with the following code.
(sharing in case anyone finds themselves in a similar situation).

aiorepl_start.py
(or main.py or however you wanna start it)
```
import asyncio
import aiorepl
import webserver

async def demo():
    await asyncio.sleep_ms(1000)
    print("async demo")

state = 20

async def task1():
    while state:
        #print("task 1")
        await asyncio.sleep_ms(500)
    print("done")

async def main():
    print("Starting tasks...")

    # Start other program tasks.
    t1 = asyncio.create_task(task1())
    webtask = asyncio.create_task(webserver.start_server())

    # Start the aiorepl task.
    repl = asyncio.create_task(aiorepl.task())

    await asyncio.gather(t1, webtask, repl)

asyncio.run(main())
```
And then in webserver.py we make sure the start_server() function is asynchronous and to await the server start and sleep.

webserver.py
```
import uasyncio
from microdot_asyncio import Microdot
from led_module import led

# Setup out GPIO constants:
led = led(23)

# setup webserver
app = Microdot()

# Make a default 'root' route:
@app.route('/')
async def hello(request):
    return 'Hello world'

# LED toggle
@app.route('/led')
async def toggle_led(request):
    print("Receive Toggle Request!")
    led.toggle()
    return "OK"

# Start the webserver
async def start_server():
    print('Starting microdot webserver app')
    try:
#        await app.start_server(port=80)
        await app.run(port=80)
        await asyncio.sleep(3)
    except:
        app.shutdown()
```

**Reply by themeadery** [reply]
@Jarshvor Does this work with webrepl also?

**Comment by Jarshvor** [comment]
> @Jarshvor Does this work with webrepl also?

I'm not sure @themeadery as I've never used the webrepl (yet).

I believe we might have to modify how the webrepl is started?
Maybe give the following line a try:
```
# Start the WEBREPL task.
    repl = asyncio.create_task(webrepl.start())
```

But I suspect the above won't work if webrepl is not coded to be asynchronous. (note the a in aiorepl)
So it very well might not work without adjustments to the webrepl code.

But I believe you can also access aiorepl over wifi(telnet) and not just serial if thats something you're after.

If you end up trying this out with webrepl do report back if possible. I'm curious. =)
Thanks!

**Reply by ned-pcs** [reply]
Yes, `webrepl` also works alongside `aiorepl` and `microdot` in async mode.

I'm starting it the usual (non-async) way:
```python
        try:
            webrepl.start()
        except:
            log.info("webrepl not started")
```
